cli_tools/diagnostics/main.go

```diff
@@ -0,0 +1,153 @@
+//  Copyright 2018 Google Inc. All Rights Reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//	  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package main
+
+import (
+	"archive/zip"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"path/filepath"
+	"time"
+)
+
+var (
+	tmpFolder, tmpFolderErr = ioutil.TempDir("", "diagnostics")
+)
+
+type logFolder struct {
+	name  string
+	files []string
+}
+
+func zipFiles(logs []logFolder, outputPath string) error {
+	newFile, err := os.Create(outputPath)
+	if err != nil {
+		return err
+	}
+	defer newFile.Close()
+	writer := zip.NewWriter(newFile)
+	defer writer.Close()
+
+	for _, folder := range logs {
+		for _, path := range folder.files {
+			file, err := os.Open(path)
+			if err != nil {
+				log.Printf("Error saving file %s with error %v\n", path, err)
+				continue
+			}
+			defer file.Close()
+
+			p := fmt.Sprintf("%s/%s", folder.name, filepath.Base(path))
+			zf, err := writer.Create(p)
+			if err != nil {
+				log.Printf("Error saving file %s with error %v\n", path, err)
+				continue
+			}
+			io.Copy(zf, file)
+		}
+	}
+	return nil
+}
+
+func uploadToSignedURL(uploadPath string, signedURL string) error {
+	client := &http.Client{
+		Timeout: 10 * time.Second,
+	}
+
+	// The signed Url gives us the actual url to upload to
+	req, err := http.NewRequest("POST", signedURL, nil)
+	if err != nil {
+		return err
+	}
+	req.Header.Set("x-goog-resumable", "start")
+	resp, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+	uploadURL := resp.Header.Get("Location")
+
+	// Upload the file
+	f, err := os.Open(uploadPath)
+	if err != nil {
+		return err
+	}
+	bodyReader, bodyWriter := io.Pipe()
+	go func() {
+		defer bodyWriter.Close()
+		defer f.Close()
+		io.Copy(bodyWriter, f)
+	}()
+
+	req, err = http.NewRequest("PUT", uploadURL, bodyReader)
+	if err != nil {
+		return err
+	}
+	_, err = client.Do(req)
+	return err
+}
+
+func moveZipFile(path string) (string, error) {
+	currDir, err := os.Getwd()
+	if err != nil {
+		return "", err
+	}
+
+	knownZipPath := filepath.Join(currDir, filepath.Base(path))
+	return knownZipPath, os.Rename(path, knownZipPath)
+}
+
+func main() {
+	if tmpFolderErr != nil {
+		log.Panic("Error creating a temporary folder. Exiting")
```

Comment:
avoid panic, for this:

var tmpFolder string

main() {
var err error
tmpFolder, err = ioutil.TempDir("", "diagnostics")
if err != nil {
  log.Fatal(err)
}
...
}

Reply:
Done